### PR TITLE
fix(ext/node) worker_threads ESM handling

### LINKED ISSUE #22840: Handling of ESM files executed as `node:worker_threads` broken
Version: 1.41.2 (release, x86_64-unknown-linux-gnu)

Right now ESM files will be [edit: sometimes] misinterpreted as CJS when executed as `node:worker_threads`.

~~It's easily reproducible by enabling the `parallel/test-worker-esmodule.js` compatibility test.~~

The reason for this defect isn't hard to find and I will soon try to contribute a simple fix to correct at least the main issue and restore the old working state.

But the mechanism used to detect the import format in `node:worker_threads` is anyway rather limited and doesn't reflect the fact, that node is also slowly changing the default format expectations and related detection mechanisms...

### LINKED ISSUE #22964: Sveltekit `deno task build` not working
```
$ deno run -A npm:create-svelte@latest my-app/
$ cd my-app/
$ deno task build
Task build vite build
vite v5.1.6 building SSR bundle for production...
"confetti" is imported from external module "@neoconfetti/svelte" but never used in "src/routes/sverdle/+page.svelte".
✓ 120 modules transformed.
error: Uncaught (in worker "") (in promise) Error: require() of ES Module /Users/divy/gh/goa/my-app/node_modules/.deno/@sveltejs+kit@2.5.3/node_modules/@sveltejs/kit/src/core/postbuild/analyse.js from /Users/divy/gh/goa/my-app not supported. Instead change the require to a dynamic import() which is available in all CommonJS modules.
    at createRequireEsmError (node:module:740:15)
    at Object.Module._extensions..js (node:module:729:13)
    at Module.load (node:module:643:32)
    at Function.Module._load (node:module:524:12)
    at Module.require (node:module:662:19)
    at require (node:module:776:16)
    at data:text/javascript,(async function() {c......d/analyse.js");})();:1:140
error: Uncaught (in promise) Error: Unhandled error. ([Object: null prototype] {
  message: 'Uncaught (in promise) Error: require() of ES Module /Users/divy/gh/goa/my-app/node_modules/.deno/@sveltejs+kit@2.5.3/node_modules/@sveltejs/kit/src/core/postbuild/analyse.js from /Users/divy/gh/goa/my-app not supported. Instead change the require to a dynamic import() which is available in all CommonJS modules.',
  fileName: 'node:module',
  lineNumber: 740,
  columnNumber: 15
})
    at NodeWorker.emit (ext:deno_node/_events.mjs:371:17)
    at NodeWorker.#handleError (node:worker_threads:155:10)
    at NodeWorker.#pollControl (node:worker_threads:175:30)
    at eventLoopTick (ext:core/01_core.js:169:7)
```

### PR DESCRIPTION
**Pull request by mash-graz**
Fixes #22840
Fixes #22964 

I still have to look for better test example, because for the very simple `parallel/test-worker-esmodule.js` and its `*.mjs` worker file it doesn't make difference if this patch gets applied, but my projects, which use `.js` files in ordinary packages, do not work anymore without this modification.

I'll leave it as draft in the meanwhile...

### CODE REVIEW

File: ext/node/polyfills/worker_threads.ts [hunk 1]
```diff
@@ -175,17 +175,17 @@ class NodeWorker extends EventEmitter {
...
         (pkg && pkg.exists && pkg.typ == "module")
       ) {
+        specifier = toFileUrl(specifier as string);
+      } else {
         const cwdFileUrl = toFileUrl(Deno.cwd());
         specifier =
           `data:text/javascript,(async function() {const { createRequire } = await import("node:module");const require = createRequire("${cwdFileUrl}");require("${specifier}");})();`;
-      } else {
-        specifier = toFileUrl(specifier as string);
       }
```
Comment (dsherret) [inline ext/node/polyfills/worker_threads.ts:189]: > Nevertheless, a more general solution for the file type detection...

There is a function, `url_to_node_resolution` on `NodeResolver`. Probably that should be used everywhere. Also, all files outside a node_modules directory are considered ESM in Deno (there is a `in_npm_package` function on `NodeResolver` to tell that).

File: ext/node/polyfills/worker_threads.ts [hunk 2]
```diff
@@ -175,17 +175,17 @@ class NodeWorker extends EventEmitter {
         // empty catch block when package json might not be present
       }
       if (
-        !(StringPrototypeEndsWith(
+        (StringPrototypeEndsWith(
           StringPrototypeToString(specifier),
           ".mjs",
         )) ||
         (pkg && pkg.exists && pkg.typ == "module")
```
Comment (dsherret) [inline ext/node/polyfills/worker_threads.ts:182]: Probably should ensure it's not a `.cjs` file.

**Review by dsherret — COMMENTED** [review]
Thanks! Let me know if you need further help on this one.

Probably a test could be added in tests/integration/npm_tests.rs

File: ext/node/polyfills/worker_threads.ts (hunk 1 above)
Comment (mash-graz) [reply ext/node/polyfills/worker_threads.ts:189]: The `url_to_node_resolution` handles only an URL to path resp. string conversion.
But it doesn't provide any file type related differentiation, if I'm not wrong.

Right now the format decision is done by looking at 1.) the file suffix and 2.) `op_require_read_closest_package_json` to check, if a `type: module` entry is present somewhere in a `package.json` file above the script.

But **the code without my patch always seemed to choose the exactly opposite consequence based on these checks!** ESM files got wrapped in `require(...)` fragments and CJS files became imported by their URL, not the other way around! (<= But please double-check this claim!)

But `.mjs` files still seemed to work all the time somehow and simply ignored this defect.
That's why the unit test doesn't report an error.

File: ext/node/polyfills/worker_threads.ts (hunk 2 above)
Comment (mash-graz) [reply ext/node/polyfills/worker_threads.ts:186]: it's now added!

**Comment by bartlomieju** [comment]
@mash-graz can you link to some of the package you said don't work without this patch? We can boil down a test case out of them.

**Comment by mash-graz** [comment]
I noticed the issue while working on this `jco` porting attempts, but this working tree is in a such a horrible chaotic state right now, that I can hardly share it.

But I'll look, if I can extract a working example till tomorrow...

It's really crazy -- I can somehow understand, what's wrong with this particular code passage, but I can not grasp, why most examples, and the `*.mjs` files in particular, nevertheless work most of the time, although the get imported in an utterly inadequate way.

**Comment by mash-graz** [comment]
o.k. -- I have prepared a practical demonstration of this issue for you:

1. checkout the `deno_demo` branch of `https://github.com/mash-graz/jco`

2. at the top level directory use this command to start the `mocha` test run for the `preview2-shim` workspace: 

```
❯ ../deno/target/debug/deno run -A npm:mocha -u tdd ./packages/preview2-shim/test/test.js

  Node.js Preview2
test stdouttest stderr    ✔ Stdio
error: Uncaught (in worker "") (in promise) Error: require() of ES Module /home/mash/Projects/jco/packages/preview2-shim/lib/io/worker-thread.js from /home/mash/Projects/jco not supported. Instead change the require to a dynamic import() which is available in all CommonJS modules.
    at createRequireEsmError (node:module:1084:15)
    at Object.Module._extensions..js (node:module:1065:13)
    at Module.load (node:module:919:32)
    at Function.Module._load (node:module:749:12)
    at Module.require (node:module:941:19)
    at require (node:module:1126:16)
    at data:text/javascript,(async function() {c......er-thread.js");})();:1:138
```

you will get this error, if you execute the unpatched `deno` git upstream version (47f318230)

but if you utilize this PR everything should work as expected and you will see something like this:
```
❯ ../deno/target/debug/deno run -A npm:mocha -u tdd ./packages/preview2-shim/test/test.js


  Node.js Preview2
test stdouttest stderr    ✔ Stdio
    1) FS read
    ✔ WASI HTTP (744ms)
    Clocks
      ✔ Wall clock
      ✔ Monotonic clock now
      ✔ Monotonic clock immediately resolved polls
      ✔ Monotonic clock subscribe duration
      ✔ Monotonic clock subscribe instant
    WASI Sockets (TCP)
      ✔ sockets.instanceNetwork() should be a singleton
      ✔ sockets.tcpCreateSocket() should throw not-supported
      ✔ tcp.bind(): should bind to a valid ipv4 address
      2) tcp.bind(): should bind to a valid ipv6 address and port=0
      ✔ tcp.bind(): should throw invalid-argument when invalid address family
      ✔ tcp.bind(): should throw invalid-state when already bound
      ✔ tcp.listen(): should listen to an ipv4 address
      3) tcp.connect(): should connect to a valid ipv4 address and port=0
    WASI Sockets (UDP)
      ✔ sockets.udpCreateSocket() should be a singleton
Warning: setTTL on UDP sockets ignored
      ✔ udp.bind(): should bind to a valid ipv4 address and port=0
Warning: setTTL on UDP sockets ignored
      ✔ udp.bind(): should bind to a valid ipv6 address and port=0
Warning: setTTL on UDP sockets ignored
      4) udp.stream(): should connect to a valid ipv4 address
Warning: setTTL on UDP sockets ignored
      ✔ udp.stream(): should connect to a valid ipv6 address (208ms)


  17 passing (2s)
  4 failing
  [...]
```

I really can't explain, why the import will not work in this case, but seems to work in many other situations. That's why I couldn't find a more suitable unit test. 

But the reported `Error: require() of ES Module` was rather clear and further debug sessions lead me finally to this section in the source code resp. the suggested changes.

**Comment by mash-graz** [comment]
I just added two additional unit tests.

But creating them, I had to notice, that exceptions generated during Worker creations, will not be caught by these tests! :(

for example, when I try to create a worker with a non-existing path specifier (here: `./cjs-file.xjs`), I will see this output:

```
[worker_threads] inheritances ...error: Uncaught (in worker "$DENO_STD_NODE_WORKER_THREAD") Module not found "file:///home/mash/Projects/deno/tests/unit_node/testdata/worker_module/cjs-file.xjs".
 ok (118ms)
```

But the test will nevertheless pass and report:

```
ok | 17 passed | 0 failed (911ms)
```

File: ext/node/polyfills/worker_threads.ts (hunk 1 above)
Comment (dsherret) [reply ext/node/polyfills/worker_threads.ts:193]: > The url_to_node_resolution handles only an URL to path resp. string conversion.
But it doesn't provide any file type related differentiation, if I'm not wrong.

It works for `file:///` urls.

**Comment by mash-graz** [comment]
I finally figured out an indirect check for correct worker script import, which will fail on errors.
It's using a combination of `error` and `message` event expectation.

This should now at least cover all the possible processing paths within the patched sequence, although the CJS test is rather weak. I don't know if it should accept `require(...)` or `__filename` instructions in the imported content?

File: ext/node/polyfills/worker_threads.ts (hunk 1 above)
Comment (mash-graz) [reply ext/node/polyfills/worker_threads.ts:193]: That's interesting! 
Which property of the URL object could be used for a more useful import type specific proceeding?

Comment (mash-graz) [reply ext/node/polyfills/worker_threads.ts:193]: @dsherret: I think, I finally understood now your `url_to_node_resolution`! :)

I'll try to use it as replacement here as well to minimize the code duplication and unexpected behavior.

**Review by bartlomieju — CHANGES_REQUESTED** [review]
Thank you for the PR @mash-graz, I'm going to mark it as "Request changes" so that we can first fix regressions in https://github.com/denoland/deno/issues/22934 and https://github.com/denoland/deno/issues/22935 to not introduce more moving pieces and potential regressions. We'll land this one as soon as we fix these two problems.

**Comment by mash-graz** [comment]
That's perfectly fine!

I try to prepare another improved/corrected version in the meanwhile. ;)

**Comment by lino-levan** [comment]
With both of the linked issues now being resolved, would this PR be up for another review?

**Comment by mash-graz** [comment]
Please wait just one or two days more! 

I'm working on a much more satisfying rewrite, which handles both input options (ULRs or path entries) equally well. That's unfortunately not case with all these earlier variants. Therefore, this temporary `draft` flag again.

**Comment by mash-graz** [comment]
It's finally ready for review (again)! :)

I spend a lot of time into input verification and early exception throwing, because initialization errors on the worker side are usually only reported as `uncaught exception` and hard to handle in unit-tests, etc.

File and Data-URL-Objects should be accepted equally than absolute and relative path strings, in compliance to the expectations defined by the `node` documentation.

Although I did my best to handle also CJS scripts sufficiently by wrapping them in helper scripts, `async` mechanisms, `Promises` and `timers` will not work in this case. If you enable this line in the unit-test, the whole process will freeze:

https://github.com/denoland/deno/blob/7a799d232729a8f2e1542e0aaeaa4acaa582c8e9/tests/unit_node/testdata/worker_module/cjs-file.cjs#L13

I hope, the changes are mostly acceptable.

And thanks again for all this other help!

**Comment by mash-graz** [comment]
sorry -- I have to debug the windows path issues tomorrow...

**Comment by bartlomieju** [comment]
@littledivy @satyarohith please review this PR, this might be overlapping with https://github.com/denoland/deno/pull/22977 you're working on.

**Comment by mash-graz** [comment]
Now it's finally passing the CI pipeline, but one important test (execution of the CJS wrapped code) is disabled on windows! :(

That's a really inconvenient situation, because fixing this necessary differences in handling of ESM vs. CJS scripts was the main goal of this PR.

How should we go on?

* Should just open another PR with a much simpler fix for the original issue as an interim solution, to quickly solve the troubles reported in #17248 and my first issue report?

* How do you value this attempt to rewrite this section in rust, instead of just using the existing TS code? All the platform specific troubles in regard to URL and file path handling increased my doubts about these attempts.

* If you are willing to merge this kind of PR, I would at least add a runtime platform check and disable the CMS support on windows.

It's very hard for me to debug the cause of these windows specific bugs. I do not have a setup to reproduce them locally and always waiting for the CI results is an enormous waste of time...

Please give me just a little bit of honest feedback, what would look as the most appreciable solution to you?

**Comment by mash-graz** [comment]
The windows bug seems to be outside of my code changes in `require()`:

```
 [worker_threads_test 011.97] thread 'worker-7' panicked at ext\node\ops\require.rs:198:57:
[worker_threads_test 011.97] called `Result::unwrap()` on an `Err` value: ()
```

I'll open another issue and Fix for this problem...
But It shouldn't be seen as a blocker for this PR, because it will happen with all other ad hoc fix attempts as well.

**Comment by littledivy** [comment]
> Should just open another PR with a much simpler fix for the original issue as an interim solution, to quickly solve the troubles reported in https://github.com/denoland/deno/issues/17248 and my first issue report?

Just tested this PR on Nuxt and Sveltekit. It solves the problem for both, let's try to get this merged.

> How do you value this attempt to rewrite this section in rust, instead of just using the existing TS code? All the platform specific troubles in regard to URL and file path handling increased my doubts about these attempts.

I like the port to Rust. Rewrites like these are welcome since they reduce the startup snapshot size.

I just noticed you enabled the tests on Windows. Does it work now?

**Comment by mash-graz** [comment]
> I just noticed you enabled the tests on Windows. Does it work now?

Yes -- it should now work on all platforms!
The CI system just stumbles over other flaky WPT problems.

I hope, it's more or less o.k. by now, but somebody should definitely take careful look on the code, because this windows related path issues nearly drove my crazy.

**Comment by mash-graz** [comment]
@bartlomieju could you please take another look at this code.

It should now work on all platforms, but im sure, many thinks could be done better.
If you find code-issues, that should be changed before merging -- no problem!

**Review by littledivy — APPROVED** [review]
LGTM

**Review by satyarohith — APPROVED** [review]
LGTM. Nice to have more operations done on the rust side.

**Comment by mash-graz** [comment]
thanks for the approvals! :)

**Pull request merged**